scripts-available/CDB_CartodbfyTable.sql

```diff
@@ -499,40 +513,62 @@ BEGIN
 
   -- Found something named right...
   IF FOUND THEN
-  
-    -- And it's an integer column...
-    IF rec.atttypid IN (20,21,23) THEN
-          
+
+    -- And it's an integer or varchar column...
+    IF rec.atttypid IN (20,21,23,1043) THEN
+
       -- And it's a unique primary key! Done!
       IF (rec.indisprimary OR rec.indisunique) AND rec.attnotnull THEN
         RAISE DEBUG 'CDB(_CDB_Has_Usable_Primary_ID): %', Format('found good ''%s''', const.pkey);
-        RETURN true;
+        SELECT true as has_usable_primary_key, text_key_column INTO idc;
+        RETURN idc;
 
-      -- Check and see if the column values are unique and not null, 
+      -- Check and see if the column values are unique and not null,
       -- if they are, we can use this column...
       ELSE
 
         -- Assume things are OK until proven otherwise...
         useable_key := true;
-      
-        BEGIN
-          sql := Format('ALTER TABLE %s ADD CONSTRAINT %s_pk PRIMARY KEY (%s)', reloid::text, const.pkey, const.pkey);
-          RAISE DEBUG 'CDB(_CDB_Has_Usable_Primary_ID): %', sql;
-          EXECUTE sql;
-          EXCEPTION      
-          -- Failed unique check...
-          WHEN unique_violation THEN
-            RAISE DEBUG 'CDB(_CDB_Has_Usable_Primary_ID): %', Format('column %s is not unique', const.pkey);
-            useable_key := false;
-          -- Failed not null check...
-          WHEN not_null_violation THEN
-            RAISE DEBUG 'CDB(_CDB_Has_Usable_Primary_ID): %', Format('column %s contains nulls', const.pkey);
-            useable_key := false;
-          -- Other fatal error
-          WHEN others THEN
-            PERFORM _CDB_Error(sql, '_CDB_Has_Usable_Primary_ID');          
-        END;
-  
+
+        -- If the column is varchar, try to cast it to integer
+        IF rec.atttypid IN (1043) THEN
```

Comment:
the point here is not to run checks on the type of the column (and btw I really hate those magic numbers) but to assume the `cartodb_id` is valid just because you're feeding cartodbfy with it.

In other words: the contract of the `CDB_CartodbfyTable` should state that _if the input table contains a column `cartodb_id` it must be a valid one: numeric, unique and not null._ as a pre-condition.